ote_sdk/ote_sdk/usecases/evaluation/anomaly_metrics.py

```diff
@@ -0,0 +1,60 @@
+""" This module contains the implementations of performance providers for multi-score anomaly metrics. """
+
+# Copyright (C) 2021-2022 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
+
+from typing import List, Optional
+
+from ote_sdk.entities.metrics import (
+    AnomalyLocalizationPerformance,
+    MetricsGroup,
+    Performance,
+    ScoreMetric,
+)
+from ote_sdk.entities.resultset import ResultSetEntity
+from ote_sdk.usecases.evaluation.averaging import MetricAverageMethod
+from ote_sdk.usecases.evaluation.dice import DiceAverage
+from ote_sdk.usecases.evaluation.f_measure import FMeasure
+from ote_sdk.usecases.evaluation.performance_provider_interface import (
+    IPerformanceProvider,
+)
+from ote_sdk.utils.dataset_utils import (
+    contains_anomalous_images,
+    split_local_global_resultset,
+)
+
+
+class AnomalyLocalizationScores(IPerformanceProvider):
+    """
+    This class provides the AnomalyLocalizationPerformance object for anomaly segmentation and anomaly detection tasks.
+    The returned performance object contains the local (pixel/bbox-level) performance metric as the main score if local
+    annotations are available. The global (image-level) performance metric is included as additional metric.
+
+    :param resultset: ResultSet that scores will be computed for
+    """
+
+    def __init__(self, resultset: ResultSetEntity):
+        self.local_score: Optional[ScoreMetric] = None
+        self.dashboard_metrics: Optional[List[MetricsGroup]] = None
+
+        global_resultset, local_resultset = split_local_global_resultset(resultset)
+
+        global_metric = FMeasure(resultset=global_resultset)
+        global_performance = global_metric.get_performance()
+        self.global_score = global_performance.score
+
+        if contains_anomalous_images(local_resultset.ground_truth_dataset):
```

Comment:
Is it allowed not to have anomalous images in dataset?

Reply:
It it is allowed, then primary metric can have different kind different from run to run (if the dataset is changed between runs).
If it is not allowed, then it always has local_metric different from None.

Reply:
In an anomaly segmentation project, the user can upload some images but not make any segmentation annotations. In this case the global (image-level label) ground truth dataset will contain anomalous images, but the local (segmentation polygon) ground truth dataset will not contain any anomalous images. When we train the model, we still want to report some performance metric to the user, so we take global score as the primary metric. The user may then annotate some images, after which we can compute the local metric. This will then be the primary metric.

We do this because we always need to report a primary metric to show in the UI. In the future, the UI will show both global and local performance based on the availability. We could then change the AnomalyLocalizationPerformance to always return the local score if needed.

Reply:
And a little bit motivation why we do this: We don't want to block the training if there is no detection/segmentation annotation available. We want to train the model as soon as the user uploads normal and abnormal images. If we forced the user to annotate to start the training, the user would be confused as we claim that anomaly tasks are unsupervised tasks.